Improve performance of BitSet.toBitMask (and .range*)

linasm opened the pull request:
This improves the performance of `BitSet.toBitMask` (and, transitively `BitSet.range*` methods) by overriding the default implementation (copying via `while` loop) with `Array.clone` in `mutable.BitSet` and `immutable.BitSet.BitSetN`.

Existing test coverage looks adequate to me.

~~Note: the results are not as convincing as I had hoped originally, so I would completely understand if this does not get approved. Still worth having a look, IMHO.~~
Update: I was advised to experiment with various GCs. Since the benchmark is allocating a lot, this indeed had a significant effect. Now, with `-XX:+UseSerialGC`, the results are more consistent and convincing: https://github.com/scala/scala/pull/8412#issuecomment-530335704

(OUTDATED, see https://github.com/scala/scala/pull/8412#issuecomment-530335704) Benchmark results (`size` is the size of underlying array of longs):
```
Original:
[info] Benchmark                           (size)  Mode  Cnt      Score     Error  Units
[info] BitSetToBitMaskBenchmark.toBitMask       1  avgt    6     11.004 ±   0.023  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask       8  avgt    6     22.476 ±   0.620  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask      64  avgt    6     83.246 ±   0.855  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask     512  avgt    6    548.364 ±  17.084  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask    4096  avgt    6   5287.501 ±  15.449  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask   32768  avgt    6  45144.393 ± 515.695  ns/op

Modified:
[info] Benchmark                           (size)  Mode  Cnt      Score     Error  Units
[info] BitSetToBitMaskBenchmark.toBitMask       1  avgt    6     12.148 ±   0.698  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask       8  avgt    6     14.945 ±   0.119  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask      64  avgt    6     78.903 ±   0.949  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask     512  avgt    6    565.565 ±   3.771  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask    4096  avgt    6   4743.584 ±  22.519  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask   32768  avgt    6  40291.784 ± 365.573  ns/op
```

Benchmark code:
```
package scala.collection.mutable

import org.openjdk.jmh.annotations._
import org.openjdk.jmh.infra._
import java.util.concurrent.TimeUnit

import scala.collection.mutable

@BenchmarkMode(Array(Mode.AverageTime))
@Fork(1)
@Threads(1)
@Warmup(iterations = 6)
@Measurement(iterations = 6)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
class BitSetToBitMaskBenchmark {

  @Param(Array("1", "8", "64", "512", "4096", "32768"))
  var size: Int = _

  var bs: mutable.BitSet = _

  @Setup(Level.Iteration) def initialize(): Unit = {
    bs = (0 until size * 64).to(mutable.BitSet)
  }

  @Benchmark def toBitMask(bh: Blackhole): Unit = {
    bh consume bs.toBitMask
  }

}
```

Review by @Ichoran

linasm committed:
Improve performance of BitSet.toBitMask

plokhotnyuk commented:
@linasm I would simplify the benchmark to `@Benchmark def toBitMask: mutable.BitSet = bs.toBitMask`

Have you tried it for `immutable.BitSet`, and with different JVMs/GCs?

linasm commented:
@plokhotnyuk thank you for the feedback!

1. I tried on both JDK 8 and 11 with very similar results (the results posted are from 11).

2. I also benchmarked `immutable.BitSet` but did not post the results because they were very similar, with exception being for the `size = 1` which uses `immutable.BitSet.BitSet1` (which is not affected by my change). I also played with overriding `BitSet1/BitSet2.toBitMask` with a plain array allocation, but saw no improvement at all (guess JVM is eliminating that `while` loop completely within those fixed size implementations), so I skipped those changes from my PR.

Following yous suggestions, I have also tried:

3. Changed benchmark method to `@Benchmark def toBitMask: Array[Long] = bs.toBitMask`: no difference.

4. Played with GCs, and indeed saw significantly better numbers with `-XX:+UseSerialGC`, also showing more consistent improvement from my changes:
```
[info] # JMH version: 1.20
[info] # VM version: JDK 11.0.4, VM 11.0.4+11
[info] # VM invoker: /Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home/bin/java
[info] # VM options: -XX:+UseSerialGC
```
Original code:
```
[info] Benchmark                           (size)  Mode  Cnt      Score     Error  Units
[info] BitSetToBitMaskBenchmark.toBitMask       1  avgt    6     11.667 ±   0.106  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask       8  avgt    6     22.271 ±   0.241  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask      64  avgt    6     60.404 ±   2.725  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask     512  avgt    6    405.337 ±  12.824  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask    4096  avgt    6   4591.375 ±  25.852  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask   32768  avgt    6  41055.380 ± 513.721  ns/op
```
Modified code:
```
[info] Benchmark                           (size)  Mode  Cnt      Score     Error  Units
[info] BitSetToBitMaskBenchmark.toBitMask       1  avgt    6     11.936 ±   0.226  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask       8  avgt    6     13.234 ±   0.243  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask      64  avgt    6     55.824 ±   1.075  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask     512  avgt    6    358.571 ±   6.671  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask    4096  avgt    6   3110.975 ±  28.581  ns/op
[info] BitSetToBitMaskBenchmark.toBitMask   32768  avgt    6  28604.277 ± 294.692  ns/op
```
